### PR TITLE
forget: Update docs for readability and append-only considerations

### PR DESCRIPTION
**Pull request by lgommans**
This is a follow-up to #3584 -- all info is there :)

### CODE REVIEW

**Comment by rawtaz** [comment]
Thank you @lgommans for your patience and for helping me fix my accidental PR messup.

So, I started reviewing your changes and fould a few things I wanted to change, for example:

- Some sentences or explanations I found a bit hard to understand some details of, and also if I try to put myself behind the eyes of a non-technical or accustomed reader I felt they needed to be made clearer.
- A couple of lines or changes I felt were too out of context or too meta for this documentation (e.g. a reference to an issue tracking re-encrypting a repository).
- Append-mode in general wasn't very well explained overall in the documentation (since before your changes), so that was one of the things I wanted to improve further.

While reviewing the thread model we also found a few things that weren't entirely accurate or complete (this had little to do with your edits, was mostly existing text), so I changed those too.

So overall I've made what comes across as a lot of editorial and grammatical changes. I hope you don't mind too much (this is why I put it in a second commit). I sincerely do appreciate that you took the time to write the text that you did, because the more I worked with this PR, the more important I felt that this addition to the documentation is.

I'd be happy to hear your thoughts if you want to review my additional changes. If you want to view them specifically you can check my commit on the Commits tab, otherwise to see the end result you can use the "Files changed" tab. Also looking forward to Michael's or anyone else's review.

**Comment by rawtaz** [comment]
@lgommans Hm, could you update https://restic-test-for-merge-request.readthedocs.io/en/latest/ so that it points to this new `forget-security` branch instead of your master branch? Then we can see the changes there.

**Comment by lgommans** [comment]
Thanks for all the work on this! That all makes a lot of sense to change and add. I kind of trust it'll be fine, but before I spot some typo after the fact I guess it makes sense that I give it a read before merging. I'll do that tomorrow, probably before ~20:00 (GMT+1).

**Comment by lgommans** [comment]
Ah I forgot to mention: the docs were built https://restic-test-for-merge-request.readthedocs.io/en/latest/

Couldn't find how to make it pull the list of branches in order to 'activate' another branch for building, so I used the element inspector to change the value for "latest" from `master` to `forget-security` (funny name, btw, not sure if that was chosen on purpose :D). It broke the admin page (showing `-----` instead of `forget-security`) but on the build page I could now make it pull that branch and build it :)

**Comment by rawtaz** [comment]
> I used the element inspector to change the value for "latest" from master to forget-security

I'm just curious, which element did you change?

> funny name, btw, not sure if that was chosen on purpose :D

Haha, it actually wasn't at all intentional, but it's really funny now that you point it out yeah!

File: doc/060_forget.rst
Comment (MichaelEischer) [inline doc/060_forget.rst:323]: Future timestamps are only ignored by the `--keep-within*` options. And the addition of the remark here causes the next sentence "This is a safety feature" to be unclear what it refers to.

File: doc/design.rst
Comment (MichaelEischer) [inline doc/design.rst:635]: s/has/have/

File: doc/design.rst
Comment (MichaelEischer) [inline doc/design.rst:669]: I'd add "only" before "be done using ...".

File: doc/060_forget.rst
Comment (MichaelEischer) [inline doc/060_forget.rst:361]: s/must have/require/

File: doc/060_forget.rst
Comment (MichaelEischer) [inline doc/060_forget.rst:360]: IMO this paragraph should also explain the append-only setup variant where a trusted host is allowed to remove data from the repository. Without such an explanation, the "trusted client" in the next paragraph is not properly defined.

File: doc/060_forget.rst
Comment (MichaelEischer) [inline doc/060_forget.rst:373]: There's a typo in the `--time` option.

File: doc/060_forget.rst
Comment (MichaelEischer) [inline doc/060_forget.rst:375]: `or a scheduled job`?

File: doc/060_forget.rst
Comment (MichaelEischer) [inline doc/060_forget.rst:377]: s/use/keep only/

**Review by MichaelEischer — COMMENTED** [review]
I've added a few small remarks. Currently, there's no explanation what a trusted client is and how it is involved in an append-only backup.

And I'm wondering whether it would be helpful to add a TL;DR box for the "Security considerations in append-only mode" section. After all the section can be summed up as "Use `--keep-within` if you want to forget data in an append-only repository" and I'm (unfortunately) not sure whether everyone manages to read the complete section.

File: doc/060_forget.rst
Comment (rawtaz) [reply doc/060_forget.rst:375]: Yeah, a `forget` run with a policy that is susceptible to the attack discussed here might be run manually by the administrator, or in an automated way by e.g. a cron job or similar. The text said "cron job" but I changed it to "scheduled job" because cron doesn't exist on Windows.

Comment (rawtaz) [reply doc/060_forget.rst:375]: Good catch, will change.

**Comment by lgommans** [comment]
> I'm just curious, which element did you change?

On <https://readthedocs.org/dashboard/restic-test-for-merge-request/advanced/> the element with ID `id_default_branch`

---

The changes look good to me! I spotted a minor typo, will add a comment there.

@rawtaz Since you said "Good catch, will change." in response to one of Michael's comments, are you already working on another commit? Just to prevent double work.

File: doc/060_forget.rst
Comment (lgommans) [inline doc/060_forget.rst:369]: `attacker's` with apostrophe

**Comment by rawtaz** [comment]
> are you already working on another commit?

I was intending to, but if you want to have a stab at it I don't mind. As long as I can add some adjustments if I feel like it :)

File: doc/060_forget.rst
Comment (lgommans) [reply doc/060_forget.rst:323]: Upon re-reading the paragraph, the future timestamp thing indeed seems kind of like a stray remark in between two connected sentences. That futures are ignored for `keep-within` is already mentioned in a `.. note::` section above. I've just removed the sentence in my upcoming commit but just treat that as a proposal.

File: doc/060_forget.rst
Comment (lgommans) [reply doc/060_forget.rst:360]: Isn't that already explained in the previous paragraph? I'm trying to come up with a short addition that makes it more clear what "trusted client" means but it all sounds redundant.

Perhaps we should instead change the phrasing of "a trusted client" into something like "a client with full access", especially since "trusted" is never used again (defining a term for single use is kind of silly). That makes the paragraph-length sentence even more convoluted though, so I've rewritten it a bit (in the upcoming commit).

**Comment by lgommans** [comment]
Interesting... because my local system time was wrong (I may or may not have been playing [worldle](https://worldle.teuteuf.fr)) it seems that GitHub 'correctly' sorts the commit into the thread by timestamp, thereby hiding it somewhere above the previous comments. Tried to amend it, and looking at the [commits tab](https://github.com/restic/restic/pull/3656/commits) it seems I succeeded in the end, so just ignore the commits in between :sweat_smile:

This commit addresses each of the comments, though in particular for "there's no explanation what a trusted client is and how it is involved in an append-only backup" (see also [this reply](https://github.com/restic/restic/pull/3656#discussion_r810680960)) I'm not sure if it's sufficiently clear now. Feel free to add commits with changes/additions/deletions, of course, or request more changes.

**Comment by lgommans** [comment]
Sorry for the delay, updated it just now, 'latest' should be up-to-date. If you have a readthedocs username for me, I can also add you as maintainer btw.

File: doc/060_forget.rst
Comment (MichaelEischer) [reply doc/060_forget.rst:363]: What about adding something like the following:
```
The usual setup is to use a separate, well-secured client that is full access to the repository and is only used to run ``prune``.
```

Comment (rawtaz) [reply doc/060_forget.rst:356]: I've incorporated your suggestion in the new text changes I just pushed, good input! I'll mark it as resolved now that @lgommans removed the reference to "trusted" earlier and I added your suggestion.

**Comment by rawtaz** [comment]
Okay, I went at it again and just polished the remaining review comments by @MichaelEischer. However, I also sidetracked a bit and updated most of the forget documentation about policies such that it is now hopefully much more readable and understandable - I think it was a bit of an unstructured mess before. Let's see what you think.

@lgommans Can you kick https://restic-test-for-merge-request.readthedocs.io/en/latest/060_forget.html#removing-snapshots-according-to-a-policy again?

**Comment by lgommans** [comment]
Docs updated!

**Review by fd0 — APPROVED** [review]
Good work, thanks!

**Comment by rawtaz** [comment]
@lgommans Thanks for your help! I didn't delete your branch just in case you want to keep it for nostalgic reasons.

**Pull request merged**
